relay-sampling/src/lib.rs

```diff
@@ -888,65 +888,31 @@ pub struct SamplingMatch {
+        self.sample_rate = new_sample_rate;
     }
 
-    /// Generates the seed used for the random number generation that is used for sampling decisions.
-    fn get_seed(
-        event: &Event,
-        dsc: Option<&DynamicSamplingContext>,
-        has_matched_trace_rule: bool,
-    ) -> Option<Uuid> {
-        // If we are in a situation in which we have matched a trace rule, we want to use the trace id
-        // as the seed for the random number generation, to guarantee the same sampling result
-        // across transactions of the same trace. The problem here is that this logic won't help
-        // in cases in which we have multiple matches of transaction and trace rules because they will
-        // effectively apply on different payloads. The transaction rule will match the event itself which differs
-        // between each component of the trace and the trace rule will match the dynamic sampling context that
-        // is the same across all components of the trace.
-        //
-        // An example of the aforementioned case:
-        // /hello -> /world -> /transaction belong to trace_id = abc
-        // * /hello has uniform rule with 0.2 sample rate which will match all the transactions of the trace
-        // * each project has a single transaction rule with different factors (2, 3, 4)
-        //
-        // 1. /hello is matched with a transaction rule with a factor of 2 and uses as seed abc -> 0.2 * 2 = 0.4 sample rate
-        // 2. /world is matched with a transaction rule with a factor of 3 and uses as seed abc -> 0.2 * 3 = 0.6 sample rate
-        // 3. /transaction is matched with a transaction rule with a factor of 4 and uses as seed abc -> 0.2 * 4 = 0.8 sample rate
-        //
-        // We can see that we have 3 different samples rates but given the same seed, the random number generated will be the same.
-        let event_id = event.id.value();
-
-        if has_matched_trace_rule {
-            if let Some(dsc) = dsc {
-                return Some(dsc.trace_id);
-            }
-        }
-
-        event_id.map(|id| id.0)
+    /// Returns an iterator of references that chains together and merges rules.
+    ///
+    /// The chaining logic will take all the non-trace rules from the project and all the trace/unsupported
+    /// rules from the root project and concatenate them.
+    pub fn merge_rules_from_configs<'a>(
+        sampling_config: &'a SamplingConfig,
+        root_sampling_config: Option<&'a SamplingConfig>,
+        default_value: &'a Vec<SamplingRule>,
+    ) -> impl Iterator<Item = &'a SamplingRule> {
+        let event_rules = sampling_config
+            .rules_v2
+            .iter()
+            .filter(|&rule| rule.ty == RuleType::Transaction || rule.ty == RuleType::Error);
+
+        let parent_rules = root_sampling_config
+            .map_or(default_value, |config| &config.rules_v2)
+            .iter()
+            .filter(|&rule| rule.ty == RuleType::Trace);
```

Comment:
Do you ever need to pass in a non-empty `default_value`? If not, I would remove that parameter and write

```suggestion
            .into_iter()
            .flat_map(|config| config.rules_v2.iter())
            .filter(|&rule| rule.ty == RuleType::Trace);
```